### ISSUE TITLE
[API DESIGN REVIEW] New Data Augmentation API

### ISSUE THREAD

**Issue by Dref360** [keras-team/keras#9092]
A new flexible, functional API!
This work is heavily borrowed from one of my project: https://github.com/Dref360/keras-transform

Please see the following document : 
https://docs.google.com/document/d/1h9g91xB56Iev_VITdhDmxwyZm-RHo8j-ZmBgW00sJio/edit?usp=sharing

**Comment by fchollet** [comment]
- I don't fully understand `mask`.
- I like the idea of applying transformations on `Sequence`
- It's nice to be able to specify the order of the transformations
- The simplest, most intuitive substrate for stateless data transformations are pure functions. Can we do something with the TF data transformation functions? https://www.tensorflow.org/api_guides/python/image#Image_Adjustments

**Comment by Dref360** [comment]
Masks are useful when you have multiple inputs or outputs

Let's say your model has 2 inputs, one image, and some text. You want to do data augmentation only on the image. You provide a mask for your entire batch so if your batch is `[[img, text],target]` you provide the mask :  `[[True,False],False]`. 

For segmentation, you need to apply data augmentation on the target as well so for a pair (X,Y) you give `[True, True]`. 


The stateless function would be more intuitive, they would return a Sequence. But it would be awkward to chain them I think? We couldn't have the `SequentialTransformer`. But I will add it on my project, since it's only a wrapper around the stateful API.

```python
random_rotation(
    random_zoom(
    random_shift(
    your_original_sequence, rng=15),zoom=(0.5, 1.2), rng=10)
```

**Comment by fchollet** [comment]
But is it a lot easier than e.g.?

```python
        class AugmentedCIFAR10(CIFAR10Seq):

            def __getitem__(self, idx):
                x, y = super(AugmentedCIFAR10).__getitem__(idx)
                x = random_shift(x, ...)
                x = random_rotation(x, ...)
                ...
                return x, y
```

(this can be done with the atomic functions in `preprocessing.image`)

**Comment by Dref360** [comment]
Yes you could do that, we would need to change those function to be deterministic in those cases so that we apply the same transformation to the input and the output. (add a param to avoid doing the randomness)

**Comment by ahundt** [comment]
What about preprocessing Layers like I describe in https://github.com/keras-team/keras/issues/6655?

```python
L = InputLabel(...)
input = Input(...)

# augmentation
x,L = Flip(axis=0)([input,L])
x,L = Flip(axis=1)([x,L])
x,L = Zoom(range=[0.5,2])([x,L])
x,L = Transform(matrix=[1,0,0,0,1,0,0,0,1], interpolation='bilinear')([x,L])
x,L = RandomCrop(fill_mode='constant', cval=0.)([x,L])

# network definition
x = Conv2D(inter_channel, (1, 1), kernel_initializer='he_uniform', padding='same', use_bias=False,
                   kernel_regularizer=l2(weight_decay))(x)

# ...snip...
model = Model(inputs=input, outputs=x, labels=L)
model.compile(Adam(), loss='categorical_crossentropy',metrics=['acc'])
# ...snip...
model.fit()
```

Reusing the layer abstraction could be a good way to meet the design goals:
- `Introduce as few new concepts as possible`
- `Use consistent naming and code patterns`

On the tf backend it can use the TF preprocessing ops, then other backends can be implemented with either python or using their own preprocessing tools. Perhaps it should be separate from the core model definition, but I think the general idea is sound.

**Comment by Dref360** [comment]
I have 2 concerns about your idea.
* We cannot multiprocess it. (TF ops are multi processed by default but what about python function) Where extending Sequences make them multiprocessed by design.
* Python function are quite slow (tf.py_func).
*  Is there something similar to tf.pyfunc in CNTK and Theano?

Also, There is already an effort to make data augmentation with layers for TF, the CNTK and Theano community doesn't seem to want to push this?

Transformer are really "Layer-like" and I would like to keep this thing. Stateless would be cool too, I suppose that we can have both.

**Comment by ahundt** [comment]
> We cannot multiprocess it. (TF ops are multi processed by default but what about python function) Where extending Sequences make them multiprocessed by design.

Why not? Layers simply define a sequence of steps to run. They can be implemented with layers if available on that backend and with multiprocessing under the hood if not.

> Is there something similar to tf.pyfunc in CNTK and Theano?

It looks like CNTK has similar functions in their `transforms`:
https://docs.microsoft.com/en-us/python/api/cntk.io.transforms?view=cntk-py-2.3

All development has stopped on Theano itself... so that would probably need to run in python.

> Python function are quite slow (tf.py_func).

Wouldn't the slowness of python be a disadvantage of implementing it with transformers, where layers have an advantage because they run on a c++ back end where supported? 

In the rare case py_func is needed, I use it and it only takes about 30-50ms per image to perform median filtering, for example.
